File: Python/247__Strobogrammatic_Number_II.py
class Solution(object):
    def findStrobogrammatic(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        dic = {"0": "0", "1":"1", "6":"9", "8":"8", "9":"6"}
        dic_mid = {"0": "0", "1":"1", "8":"8"}
        first_half = n / 2
        self.result = []
        def helper(count, first_half, second_half):
            if count > n:
                return
            if count == n:
                self.result += [first_half + second_half]
                return
            if count + 1 == n:
                self.result += [first_half + char + second_half for char in dic_mid]
                return
            for char in dic:
                if char == "0" and len(first_half) == 0:
                    continue
                helper(count + 2, first_half + char, dic[char] + second_half)
        helper(0, "", "")
        return self.result

File: Java/247__Strobogrammatic_Number_II.java
class Solution {
    public List<String> findStrobogrammatic(int n) {
        List<String> result = new ArrayList<>();
        helper(0, n, result, "", "");
        return result;
    }

    void helper(int index, int n, List<String> result, String left, String right) {
        if (index * 2 == n) {
            result.add(left + right);
            return;
        } else if (index * 2 + 1 == n) {
            result.add(left + "1" + right);
            result.add(left + "8" + right);
            result.add(left + "0" + right);
            return;
        }
        helper(index + 1, n, result, left + "1", "1" + right);
        helper(index + 1, n, result, left + "8", "8" + right);
        helper(index + 1, n, result, left + "6", "9" + right);
        helper(index + 1, n, result, left + "9", "6" + right);
        if (index !=0)
             helper(index + 1, n, result, left + "0", "0"+ right);
    }
}
